### PR TITLE
Elsi sparsity mask

### PR DESCRIPTION
**Pull request by bhourahine**
Fixes hermitian symmetry for parallel Pauli hamiltonians but also uses ELSI H and S pattern for dense matrices.

### CODE REVIEW

File: prog/dftb+/lib_dftbplus/mainio.F90 [hunk 1]
```diff
@@ -2786,7 +2787,8 @@ subroutine writeDetailedOut1(fd, iDistribFn, nGeoSteps, iGeoStep, tMD, tDerivs,
       end if
       if (any(electronicSolver%iSolver == [electronicSolverTypes%qr,&
           & electronicSolverTypes%divideandconquer, electronicSolverTypes%relativelyrobust,&
-          & electronicSolverTypes%elpa, electronicSolverTypes%elpadm])) then
+          & electronicSolverTypes%elpa, electronicSolverTypes%elpadm,&
+          & electronicSolverTypes%pexsi, electronicSolverTypes%ntpoly])) then
```
Comment (aradi) [inline prog/dftb+/lib_dftbplus/mainio.F90:2791]: What about introducing an `electronicSolver%providesFreeEnergy` attribute similar to '%providesEigenvalues`?

File: prog/dftb+/lib_dftbplus/mainio.F90 [hunk 2]
```diff
@@ -3823,9 +3830,9 @@ subroutine printEnergies(energy, TS, electronicSolver, tDefinedFreeE)
     if (electronicSolver%providesEigenvals) then
       write(stdOut, format2U) "Extrapolated to 0", energy%Ezero, "H", Hartree__eV * energy%Ezero,&
           & "eV"
-      write(stdOut, format2U) "Total Mermin free energy", energy%EMermin, "H",&
-          & Hartree__eV * energy%EMermin, "eV"
     end if
+    write(stdOut, format2U) "Total Mermin free energy", energy%EMermin, "H",&
```
Comment (aradi) [inline prog/dftb+/lib_dftbplus/mainio.F90:3834]: Hier we could also make an `if (electronicSolver%providesFreeEnergy) then` block...

File: prog/dftb+/lib_elecsolvers/elsisolver.F90 [hunk 3]
```diff
@@ -455,6 +473,42 @@ subroutine TElsiSolver_init(this, inp, env, nBasisFn, nEl, iDistribFn, nSpin, iS
   end subroutine TElsiSolver_init
 
 
+  !> Checks for supported value of ELSI api, 2.6.0 release tag being the currently required version
+  !> for DFTB+ can use at the moment
+  subroutine supportedVersionNumber(major, minor, patch, dateStamp)
```
Comment (aradi) [inline prog/dftb+/lib_elecsolvers/elsisolver.F90:478]: Could you please update it to 2.6.1 being the last official release before the 20.1 release.

File: prog/dftb+/lib_elecsolvers/elsisolver.F90 [hunk 4]
```diff
@@ -490,7 +490,7 @@ subroutine supportedVersionNumber(major, minor, patch, dateStamp)
       isSupported = .false.
     elseif (major == 2 .and. minor < 6) then
       isSupported = .false.
-    elseif (major == 2 .and. minor == 6 .and. any(patch == [0,1])) then
+    elseif (major == 2 .and. minor == 6 .and. .not.any(patch == [0,1])) then
```
Comment (awvwgk) [inline prog/dftb+/lib_elecsolvers/elsisolver.F90]: `.not.any(patch == [0,1])` === `all(patch /= [0,1])`

Comment (bhourahine) [reply prog/dftb+/lib_elecsolvers/elsisolver.F90]: Yes, it is

**Comment by aradi** [comment]
@bhourahine Could you get it up to date with master? Then it can be merged.

**Pull request merged**